src/psyclone/tests/psyir/frontend/fparser2_select_type_test.py

```diff
@@ -0,0 +1,758 @@
+        f"    if (type_string == 'character_star') then\n"
+        f"      character_type = ptr_character_star\n"
+        f"    end if\n\n"
+        f"  end subroutine select_type\n\n"
+        f"end module select_mod\n").lower()
+    psyir = fortran_reader.psyir_from_source(code)
+    result = fortran_writer(psyir).lower()
+    assert result == expected
+    assert Compile(tmpdir).string_compiles(code)
+    assert Compile(tmpdir).string_compiles(result)
+
+
+@pytest.mark.parametrize(
+    "pre_attribute, post_attribute",
+    (["TARGET", "TARGET"],
+     ["POINTER", "POINTER"],
+     ["ALLOCATABLE", "ALLOCATABLE, TARGET"]))
+def test_class_target(
+        fortran_reader, fortran_writer, tmpdir, pre_attribute, post_attribute):
+    '''Test that the type_selector remains unchanged if it already has the
+    target or pointer attribute and that target is added if there is a
+    pre-existing unrelated atttribute.
+
+    '''
+    code = (
+        f"module select_mod\n"
+        f"  contains\n"
+        f"  subroutine select_type(type_selector, character_type)\n"
+        f"    class(*), {pre_attribute} :: type_selector\n"
+        f"    character(len=*) :: character_type\n"
+        f"    select type(type_selector)\n"
+        f"      type is (character(len = *))\n"
+        f"        character_type = type_selector\n"
+        f"    end select\n"
+        f"  end subroutine select_type\n"
+        f"end module select_mod\n")
+    expected = (
+        f"module select_mod\n"
+        f"  implicit none\n"
+        f"  public\n\n"
+        f"  contains\n"
+        f"  subroutine select_type(type_selector, character_type)\n"
+        f"    CLASS(*), {post_attribute} :: type_selector\n"
+        f"    CHARACTER(LEN = *) :: character_type\n"
+        f"    character(256) :: type_string\n"
+        f"    CHARACTER(LEN=256), pointer :: ptr_CHARACTER_star => null()\n\n"
+        f"    type_string = ''\n"
+        f"    SELECT TYPE(type_selector)\n"
+        f"      TYPE IS (CHARACTER(LEN = *))\n"
+        f"      type_string = \"character_star\"\n"
+        f"      ptr_CHARACTER_star => type_selector\n"
+        f"    END SELECT\n"
+        f"    if (type_string == 'character_star') then\n"
+        f"      character_type = ptr_CHARACTER_star\n"
+        f"    end if\n\n"
+        f"  end subroutine select_type\n\n"
+        f"end module select_mod\n").lower()
+    psyir = fortran_reader.psyir_from_source(code)
+    if pre_attribute == "ALLOCATABLE":
+        # If the argument doesn't already have a POINTER/TARGET attribute
+        # then we can't handle it and should get a CodeBlock.
+        routine = psyir.walk(Routine)[0]
+        assert len(routine.children) == 1
+        assert isinstance(routine.children[0], CodeBlock)
+    else:
+        result = fortran_writer(psyir).lower()
+        assert result == expected
+        assert Compile(tmpdir).string_compiles(code)
+        assert Compile(tmpdir).string_compiles(result)
```

Comment:
Thanks for all the tests. One doubt that I still have is if the replacement from references to pointers is an issue when going down the call stack if the arguments don't have pointer attribute. For instance:
```
class(*) :: type_selector
SELECT TYPE (type_selector)
    TYPE IS (INTEGER)
        call my_sub(type_selector)
...
subroutine my_sub(value)
    integer :: value
```
 
I understand it would be replaced by: `call my_sub(ptr_integer)`, would this match the routine signature?

Reply:
Yes, I think that's fine. We essentially did this when we changed the LFRic PSy-layer generation to do:

    fld_ptr => my_fld_proxy%data
    ...
    call my_kernel(fld_ptr,...)

and everything was OK. I guess it's quite safe as Fortran is always pass-by-reference.